### ISSUE TITLE
Modularize lang.def 

### ISSUE THREAD

**Issue by PeterTh** [insieme/insieme#10]
Pull out non-essential IR constructs from lang.def and move to separate extensions.

**Comment by HerbertJordan** [comment]
The one processing this task will have to consider that at its current state the parser is by default only including constructs from the lang-def file as pre-defined constructs. By splitting it, some IR code fragments may not be parseable anymore.

As part of this task it is probably necessary to integrate some sort of include-mechanism into the parser to load definitions from various language extensions -- i can provide hints on those if required.

**Comment by PeterTh** [comment]
You are out of date, we already have an include mechanism :P
(see https://github.com/insieme/insieme/commit/e5a3a51bef608e503f2ba377d3dfcf6c810abd48 and related)

**Comment by HerbertJordan** [comment]
Nice. It seems I'm really a bit behind. I will have to go through the commit log one day ...

**Comment by HerbertJordan** [comment]
I studied the current solution and i have one question (which may also be already resolved?):
Isn't the central extension registry causing dependencies from the core to every extension. If language extensions are only defined for sub-modules, e.g. the backend, this is introducing cyclic module dependencies. Also, if extensions are only defined for unit tests, those maybe shouldn't show up in the global registry. Has this been resolved?

**Comment by PeterTh** [comment]
No, this is a remaining issue. The central problem is that C++ doesn't have reflection, so the only way I could see to get from a string (the extension name) to a class instance (at runtime) - which you need for the parser - was to manually create a mapping.

If anyone has any idea on how to resolve this please tell.

**Comment by HerbertJordan** [comment]
One Idea I was thinking of was using a list of template parameters. For instance the use case
```
auto node = builder.parse(" -- some code --");
```
could be modified to
```
auto node = builder.parse<VectorExtension,TestExtension>(" -- some code -- ");
```
We could define extensions that are mere collections of other extensions. We could also work with some functor as a return value and create something like
```
auto node = Parser
    .include<VectorExtension>()
    .include<TestExtension>()
    .in(" --- some code -- ")
    .as<ExpressionPtr>()
```
What do you think about something like this?

**Comment by PeterTh** [comment]
This works for code using the parser, and looks neat, but the issue requiring a {string : extension} mapping is if you want to include an extension in written Inspire code (not C++).

**Comment by HerbertJordan** [comment]
One idea would be to define language extensions within actual IR source files using let bindings which can then be imported into other source files. We could realize this importing using the C pre-processor, thereby also inheriting MACRO support.

For instance, we could have a file `vectors.ir` containing definitions like
```
let vector.init = ....
let vector.reduce = ...
```
and in a user file you write
```
#include <vector>

int<4> main() {
   -- use your vectors constructs here --
}
```
This, however, introduces the problem of two definition points of language constructs -- the source file and the C++ language extension which should offer a corresponding C++ API. We could drop the latter, but this will cause some serious refactoring and lead to a more inconvenient utilization of language extensions in the compiler code. 

Maybe keeping files synchronized is a case for a ruby script generating one out of the other?

**Comment by HerbertJordan** [comment]
A side-effect would be a clean, artifact free definition of IR language constructs in separate source files. Not the worst outcome I can imagine ...
